File: framework/src/main/java/org/synsystems/onlypass/framework/events/Event.java
package org.synsystems.onlypass.framework.events;

import android.support.annotation.NonNull;

import io.reactivex.Completable;
import io.reactivex.Single;

/**
 * Persistent record of an event that can happen during app execution.
 */
public interface Event {
  /**
   * @return a new single that emits true if this event has occurred, false otherwise
   */
  @NonNull
  public Single<Boolean> hasOccurred();

  /**
   * @return a new completable that records the event as having occurred
   */
  @NonNull
  public Completable recordOccurrence();

  /**
   * @return a new completable that clears any record of the event
   */
  @NonNull
  public Completable clearRecord();
}

File: framework/src/main/kotlin/org/synsystems/onlypass/framework/events/Event.kt
package org.synsystems.onlypass.framework.events

import io.reactivex.Completable
import io.reactivex.Single

/**
 * Persistent record of an event that can happen during app execution.
 */
interface Event {
  /**
   * Whether or not the event has occurred.
   */
  fun hasOccurred(): Single<Boolean>

  /**
   * Records the event as having occurred.
   */
  fun recordOccurrence(): Completable

  /**
   * Clears any record of the event. Safe to call if the event hasn't been recorded.
   */
  fun clearRecord(): Completable
}